### PR TITLE
Week 2 assessment ari

### PR DESCRIPTION
**Pull request by aribeltz**
# Submitting Week 2 Assessment

### ✅ Completed Files

Select the following as confirmation that the files in this repository are complete.

- [ ✅ ] Coding Challenges
- [ ✅ ] Coding Concepts
- [ ✅ ] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:

I don't think I have any specific feedback to ask for this time, I would just love to know overall what you think of my code/pseudocode/interview answers.

Also, I'd like to know what my assignment grade is if possible - Thank you!

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -14,46 +14,272 @@
...
 
+/*
+---- TEST TEMPLATE ----
+describe("", () => {
+  it(``, () => {
+    expect().toEqual()
+  })
+})
+--------
+// I changed the quotation marks inside of the it function to a template literal, just to reduce the headache of having to change quotation marks inside of it. It seems to work just fine, but please let me know if doing so would be a bad convention!
```
Comment (elysemontano) [inline code-challenges.test.js:32]: Not sure this is necessarily bad convention, just not the standard convention.

File: code-challenges.test.js [hunk 2]
```diff
@@ -14,46 +14,272 @@
...
+
+---- PSEUDOCODE TEMPLATE ----
```
Comment (elysemontano) [inline code-challenges.test.js:34]: Nice, I love this idea!!  This would be great to share with the class as well 👏

File: code-challenges.test.js [hunk 3]
```diff
@@ -14,46 +14,272 @@
...
+    expect(isObjectEven(object3)).toEqual("-7 is not divisible by three");
+  });
+});
+
+// RED:
+// Good fail, error message:  ReferenceError: isObjectEven is not defined
 
 // b) Create the function that makes the test pass.
+/*
+// Before I get into coding this, I need to figure out how to make a function that's able to reliably access the key:value pairs inside of an object. After doing some research, I've discovered some methods that work specifically on objects. The one I'm interested in is.values(). Using this, I should be able to convert just the value of the key:value pairs into an array, which I can then use something like .filter() on for my comparison. To confirm this, I'm going to run a few tests of my own before I get into building the function.
```
Comment (elysemontano) [inline code-challenges.test.js:80]: Thank you for adding this here. I am happy to see your process and thorough research.  Keep in mind, that you can also access the value of an object by using dot notation.  Ex: object1.number would give you 15.

File: code-challenges.test.js [hunk 4]
```diff
@@ -14,46 +14,272 @@
...
+//     return `${object.number} is divisible by three`;
+//   } else if (object.number % 3 !== 0) {
+//     return `${object.number} is not divisible by three`;
+//   } else {
+//     return "Oops! We've hit a snag.";
+//   }
+// };
+// Result: Test Suites: 1 passed, 1 total
+
+// I'm happy with the result of my original function, so that's the one I'm going to leave un-commented. I think it's following the theory of making functions as reusable as possible, so I'm happy I did the research to discover object methods!
```
Comment (elysemontano) [inline code-challenges.test.js:137]: Awesome!  It is always good to try multiple ways as this will help you discover tools that may be useful in the future.  Just note that Object.values will give you ALL of your values in the object which has some great use cases, but for this case might give unexpected results if other keys are in the object with a different data type.

File: code-challenges.test.js [hunk 5]
```diff
@@ -14,46 +14,272 @@
...
+// function capitalizeWords(arr) {
+//   return arr.map(element => {
+//     return element.charAt(0).toUpperCase() + element.slice(1).toLowerCase();
+//   });
+// }
+So, after seeing this, I realized I could use the addition operator to simply add the first letter to the latter half of the word by applying .charAt(0).toUpperCase() to the first letter, then .slice(1) to remove that letter and return the rest of the array. I also chose to leave in .toLowerCase(), because I want to stick with the idea of making the function as reusable as possible. We know the strings that we were given are all lowercase, but what if we put in strings that were all uppercase? Including .toLowerCase() will ensure that no matter what string we give this function, it will ALWAYS return that string with ONLY the first letter capitalized. I included that scenario in my test, just to confirm, and the function still passed!
```
Comment (elysemontano) [inline code-challenges.test.js:220]: Excellent pseudocode and I really like that you added an edge case into your test!  Thinking outside the box and how a user _might_ break something and accounting for that in your code is going to be a very valuable skill!

File: coding-concepts.js [hunk 6]
```diff
@@ -7,62 +7,90 @@
...
+const learnStudent = new Learn("George");
+console.log(learnStudent);
+// Output after running: Learn { student: 'George', cohort: 'Foxtrot', year: 2022 }
+
+// a) Your answer: This should log a new object, which will contain the following key:value pairs:
+// { student:"George", cohort:"Foxtrot", year: 2022 }
 
-// a) Your answer:
-// b) Verify and explain:
+// b) Verify and explain: I was correct about it logging a new object, I just forgot to include the class name at the beginning. This is happening because we're creating a new instance of the class, which is being saved to the variable learnStudent. When we pass the string "George" into the argument, it takes the form of the "name" parameter, then gets assigned all of the key:value pairs that are associated with that class.
```
Comment (elysemontano) [inline coding-concepts.js:96]: Excellent work in this section!

File: interview-questions.md [hunk 7]
```diff
@@ -8,50 +8,50 @@ Then, research the question to expand on your answer. Even if you feel you have
...
-Researched answer:
+Researched answer: In addition to my previous response, it's important to note that objects are also a data type in the Javascript language. They are considered non-primitive data types, along with classes.
 
 6. STRETCH: What is hoisting in JavaScript?
 
-Your answer:
+Your answer: That's a great question! I'm not personally familiar with the term hoisting in regards to Javascript, but based on the word "hoisting", I would infer that it has something to do with bringing something to the top of your code. I'll definitely look into it, and get back to you with a better answer!
 
-Researched answer:
+Researched answer: The process of hoisting in Javascript refers to the practice of moving all of your declarations to the top of their scope. This can be very beneficial, especially variable declaration. Although it may seem obvious, it's a very good idea to make sure that all of your variables have been declared BEFORE running your code.
```
Comment (elysemontano) [inline interview-questions.md:43]: Crushed this section!  Your answers are really thorough and I really like how you extended in the researched answers to bring new depth to the answer.  Great work!

**Review by elysemontano — APPROVED** [review]
You crushed this assessment Ari!  I really appreciate how thorough each answer is, I can tell you invest a lot of time into the process.  Your code and tests look really clean and it is really great you are adding in edge cases as well. Your understanding of the concepts show a competent skill level.  Keep up the great work!
